### ISSUE TITLE
extracted null value when range condition query 

### ISSUE THREAD

**Issue by sophie-jeong** [scylladb/scylla#8122]
*Installation details*
Scylla version (or git commit hash): ScyllaDB 4.1.3
OS (RHEL/CentOS/Ubuntu/AWS AMI):CentOS 7

Hello , 
as we know that the ScyllaDB can not supported to inquiry the null value.  
but null data can extracted  through the below my query. 

please think about it.
if we inquiry using new data column with the conditional after we added new data column  , 
we probably don't want the old data filled with nulls to come out with the results of your range query.

but here is the result , we can see the null data from the inquiry with the range conditional.
I think this is not the nosql policy , so I upload it as a issue in here.

******** ScyllaDB ********
test_keyspace> select * from test_tbl;
 id | col1
----+---------
  1 |   user1
  2 | user2
  3 | user3
(3 rows)
test_keyspace> alter table test_tbl add expired timestamp;
test_keyspace> select * from test_tbl;
 id | col1    | expired
----+---------+---------
  1 |   user1 |    null
  2 | user2 |    null
  3 | user3 |    null
(3 rows)
test_keyspace> insert into test_tbl (id, col1, expired) values (4,'user5', '2024-05-01 12:00:00');
test_keyspace> insert into test_tbl (id, col1, expired) values (5,'user4', '1990-05-10 03:30:00');
test_keyspace> select * from test_tbl;
 
 id | col1    | expired
----+---------+---------------------------------
  5 |   user4 | 1990-05-09 18:30:00.000000+0000
  1 |   user1 |                            null
  2 | user2 |                            null
  4 |  user5 | 2024-05-01 03:00:00.000000+0000
  3 | user3 |                            null
(5 rows)
test_keyspace> select * from test_tbl WHERE expired < dateof(now()) ALLOW FILTERING;          
 id | col1    | expired
----+---------+---------------------------------
  5 |   user4 | 1990-05-09 18:30:00.000000+0000
  1 |   user1 |                            null
  2 | user2 |                            null
  3 | user3 |                            null
---MORE--
(4 rows)
test_keyspace> select * from test_tbl WHERE expired > dateof(now()) ALLOW FILTERING;
 
 id | col1   | expired
----+--------+---------------------------------
  4 | user5 | 2024-05-01 03:00:00.000000+0000
 
---MORE---
(1 rows)

******** PostgreSQL ********
postgres=# select * from test1;
 id | col1 |        col2        
----+------+---------------------
  1 | a    | 2021-02-19 12:00:00
  2 | b    | 1992-02-19 12:00:00
  3 | c    |
  4 | d    | 2024-02-19 12:00:00

postgres=# select * from test1 where col2 < now();                 
 id | col1 |        col2        
----+------+---------------------
  1 | a    | 2021-02-19 12:00:00
  2 | b    | 1992-02-19 12:00:00
 
postgres=# select * from test1 where col2 > now();
 id | col1 |        col2        
----+------+---------------------
  4 | d    | 2024-02-19 12:00:00

**Comment by tzach** [comment]
I can confirm Scylla is different than Apache Cassandra 3.11.10

```
CREATE TABLE test_tbl (id int,col1 text, PRIMARY KEY(id));
qlsh:mykeyspace> select * from test_tbl;

 id | col1  | expired
----+-------+---------------------------------
  5 | user4 | 1990-05-10 03:30:00.000000+0000
  1 | user1 |                            null
  2 | user2 |                            null
  4 | user5 | 2024-05-01 12:00:00.000000+0000
  3 | user3 |                            null

(5 rows)
cqlsh:mykeyspace> select * from test_tbl WHERE expired < dateof(now()) ALLOW FILTERING;

 id | col1  | expired
----+-------+---------------------------------
  5 | user4 | 1990-05-10 03:30:00.000000+0000

(1 rows)
cqlsh:mykeyspace> select * from test_tbl WHERE expired > dateof(now()) ALLOW FILTERING;

 id | col1  | expired
----+-------+---------------------------------
  4 | user5 | 2024-05-01 12:00:00.000000+0000
```

**Comment by nyh** [comment]
I have a vague recollection that @avikivity and @dekimir had a long philosophical argument about this and what it should do and what it does in Cassandra and in SQL. I just don't remember what the conclusions were :-)
In any case, if we differ from Cassandra in this, I want to write a cql-pytest test that demonstrates this. I'll do it now.

**Comment by nyh** [comment]
I tried to reproduce this bug on a recent Scylla, and then realized that I can't - it was fixed. This is a duplicate of https://github.com/scylladb/scylla/issues/6295. The fix is c69075bbef9a372bb4369a881be840c6f7c38684, and it appeared in the beginning of Scylla 4.3 (and of course, later), but was not backported to the 4.1 series that you are using, and will not be backported because 4.1 is no longer support. You should consider upgrading to a more recent version (the 4.3 series already has a lot of experience, and 4.4 should be out soon).

**Comment by dekimir** [comment]
> I have a vague recollection that @avikivity and @dekimir had a long philosophical argument about this and what it should do and what it does in Cassandra and in SQL. I just don't remember what the conclusions were :-)

Handy reminder in https://github.com/scylladb/scylla/issues/7872
:smile:
